### PR TITLE
feat(tests): test highlights

### PR DESCRIPTION
**Pull request by nawordar**
Every time I change something in D highlights I am afraid that I will break highlight in some other place. D is a quite complex language, so testing manually can be cumbersome.

This PR is in no way complete, so I am creating a draft. I would also like to know what other think about such tests.

~For now I am stuck with getting a syntax tree from an opened file in a buffer. `vim.treesitter.get_parser(0):parse():root()` crashes with `attempt to call method 'root' (a nil value)`. Maybe someone knows a better way?~

### TODO
- [x] Find a way to get parser tree
- [ ] Fix duplicated results from `get_highlight_captures`
- [ ] Reduce duplicated code between `tests/indent` and `tests/highlight`
- [ ] Write actual tests

### CODE REVIEW

**Comment by nawordar** [comment]
I created a command that is almost identical and yet, it doesn't return duplicated results as `get_highlight_captures` does:
```lua
vimp.map_command('Captures', function()
  local captures = {}
  local ft = vim.bo.filetype
  local lang = require('nvim-treesitter.parsers').ft_to_lang(ft)
  local parser = vim.treesitter.get_parser(0, lang)
  local root = parser:parse()[1]:root()
  local query = vim.treesitter.query.get_query(lang, 'highlights')

  for id, node in query:iter_captures(root, 0, 0, -1) do
    local begin_row, begin_column, end_row, end_column = node:range()

    table.insert(captures, {
      name = query.captures[id],
      node = {
        range = {
          begin_row = begin_row,
          begin_column = begin_column,
          end_row = end_row,
          end_column = end_column,
        },
      },
    })
  end

  print(vim.inspect(captures))
end)
```
I have no idea why the results are duplicated and I don't want to add dirty workarounds like skipping every second result

File: tests/highlight/common.lua [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    local capture = {}
+
+    capture.name = query.captures[id]
+    capture.begin_row, capture.begin_column, capture.end_row, capture.end_column = node:range()
+
+    table.insert(captures, capture)
+  end
+
+  return captures
+end
```
[the remark this answers is not in the document]
Comment (stsewd) [reply tests/highlight/common.lua:34]: @nawordar the duplicated `@namespace` can be solved with https://github.com/nvim-treesitter/nvim-treesitter/pull/1954. I don't see the `include` node being duplicated with the example from

```d
 import std.stdio: writeln;
```

[the remark this answers is not in the document]
Comment (nawordar) [reply tests/highlight/common.lua:34]: I get duplicated nodes only when running `scripts/run_tests.sh`. `:TSPlaygroundToggle` and `:Captures` I wrote in a different comment, don't show duplicated nodes.

I've just noticed that commenting the first line in `scripts/minimal_init.lua` fixes that problem. It looks like there are two nvim-treesitters running at the same time

**Comment by theHamsta** [comment]
It would maybe be nice to use the same test format as upstream tree-sitter does: https://github.com/tree-sitter/tree-sitter/blob/master/cli/src/query_testing.rs#L27-L124 Either create a Rust CLI that parses the assertions to JSON to be accessible from Lua or create a Lua module in Rust. I really like the idea. Maybe I have time for it on Sunday

**Comment by theHamsta** [comment]
A first draft: https://github.com/theHamsta/highlight-assertions It will parse the tree-sitter test format and output it as JSON (position/expected)

**Comment by nawordar** [comment]
> It would maybe be nice to use the same test format as upstream tree-sitter does: https://github.com/tree-sitter/tree-sitter/blob/master/cli/src/query_testing.rs#L27-L124 Either create a Rust CLI that parses the assertions to JSON to be accessible from Lua or create a Lua module in Rust. I really like the idea. Maybe I have time for it on Sunday

I was thinking about it too. Making it as a separate application is a great idea, as it can help other users too.

Honestly, I kind of gave up for now with this pull request because of the problems with reliably running tests. As I wrote in a different comment, that commenting the first line in `minimal_init.lua` fixes the problem with duplicated output, but as far as I understand, this way it tests `nvim-treesitter` installed via packer, not the version that I actually want to test.

**Comment by theHamsta** [comment]
I try to finish a testing system. The tree-sitter format could help to not re-invent the wheel while we can still not only test the query system but also how Neovim highlights the files.

So you want to close this PR for now?

**Comment by nawordar** [comment]
OK, I will close it for now. Maybe I will pick it up later, when you're done with the testing system.

**Pull request closed without merging**